File: apple/LiftKit/Sources/LiftKit/TodayTotals.swift
import Foundation

/// Calories and the three macros the rectangular complication shows.
///
/// `WatchFood` also carries fibre; it is deliberately absent here. The slot
/// holds three lines and fibre is the least glanceable of the four.
public struct NutritionTotals: Equatable, Sendable {
    public var kcal: Double
    public var protein: Double
    public var carbs: Double
    public var fat: Double

    public static let zero = NutritionTotals(kcal: 0, protein: 0, carbs: 0, fat: 0)

    public init(kcal: Double, protein: Double, carbs: Double, fat: Double) {
        self.kcal = kcal
        self.protein = protein
        self.carbs = carbs
        self.fat = fat
    }
}

public enum TodayTotals {

    /// Folds every entry logged on the same **local** day as `reference`.
    ///
    /// `StandaloneFoodLog.entries` hands back the entire retained log, so this
    /// fold is the only thing narrowing it to today.
    ///
    /// Day membership is `Calendar.isDate(_:inSameDayAs:)` — never arithmetic
    /// on a timestamp. A UTC boundary misfiles an evening log as tomorrow,
    /// and 86,400-second arithmetic repeats or skips a day across a DST
    /// change.
    public static func totals(from entries: [LoggedFood],
                              on reference: Date = Date(),
                              calendar: Calendar = .current) -> NutritionTotals {
        entries
            .filter { calendar.isDate($0.loggedAt, inSameDayAs: reference) }
            .reduce(into: NutritionTotals.zero) { totals, entry in
                let per100g = entry.grams / 100
                totals.kcal += entry.food.kcal * per100g
                totals.protein += entry.food.protein * per100g
                totals.carbs += entry.food.carbs * per100g
                totals.fat += entry.food.fat * per100g
            }
    }
}

File: android/liftkit/src/main/kotlin/com/dugcanlift/liftkit/TodayTotals.kt
package com.dugcanlift.liftkit

import java.time.Instant
import java.time.ZoneId

/**
 * Calories and the three macros the face shows. Fibre is tracked by [WatchFood]
 * but deliberately absent: the face has room for three bars.
 *
 * Twin of apple/LiftKit's `NutritionTotals`. Keep the two in step.
 */
data class NutritionTotals(
    val kcal: Double,
    val protein: Double,
    val carbs: Double,
    val fat: Double
) {
    companion object {
        val ZERO = NutritionTotals(0.0, 0.0, 0.0, 0.0)
    }
}

object TodayTotals {

    /**
     * Folds every entry logged on the same **local** day as [reference].
     *
     * Day membership is a `LocalDate` comparison in [zone], never arithmetic on
     * the epoch seconds. A UTC boundary misfiles an evening log as tomorrow,
     * and 86,400-second arithmetic repeats or skips a day across a DST change.
     */
    fun totals(
        entries: List<LoggedFood>,
        reference: Instant = Instant.now(),
        zone: ZoneId = ZoneId.systemDefault()
    ): NutritionTotals {
        val today = reference.atZone(zone).toLocalDate()
        return entries
            .filter { Instant.ofEpochSecond(it.loggedAtEpochSeconds).atZone(zone).toLocalDate() == today }
            .fold(NutritionTotals.ZERO) { running, entry ->
                val per100g = entry.grams / 100
                NutritionTotals(
                    kcal = running.kcal + entry.food.kcal * per100g,
                    protein = running.protein + entry.food.protein * per100g,
                    carbs = running.carbs + entry.food.carbs * per100g,
                    fat = running.fat + entry.food.fat * per100g
                )
            }
    }
}

/**
 * The daily targets the face measures against. Defaults match LIFT iOS's own
 * `@AppStorage` values so a watch that has never been told otherwise shows
 * real targets rather than zeroes.
 */
data class Goals(
    val calories: Double,
    val protein: Double,
    val carbs: Double,
    val fat: Double,
    val steps: Double
) {
    companion object {
        val FALLBACK = Goals(
            calories = 1748.0,
            protein = 160.0,
            carbs = 167.0,
            fat = 49.0,
            steps = 10_000.0
        )
    }
}
